feat: TAO staking

UrbanWill opened the pull request:
# Description

Implements TAO staking

##  ⚠️ Warning 
- Depends on https://github.com/TalismanSociety/talisman-config/pull/7
- Add `TAOSTATS_API_KEY` to .env

## Techical 

- Renamed folder/components from Nomination pool bond specific to generic bonding names, eg: `NomPoolBondModal` -> `BondModal`
- Created bittensor and nompool folders inside a new hooks folder for organization purposes
- Created `useGetStakeInfo` and `useGetUnbondInfo` hooks to support the wizard hooks. These hooks are responsible for returning all the information that is specific to an Asset that is being staked/unbonded.
- Removed fake fee estimate, use a fee estimation that defaults to `minJoinBond` if the input amount is 0 or less than min for that pool, disable form submission if `planks` is < 0n


## TODO's

- [x] Hooks for fetching TAO delegator data
- [x] Pop up UI for selecting delegator
- [x] Dashboard UI for selecting a delegator
- [x] Refactor dashboard asset view to display the amount stake per Delegator
- [x] Hook to track the block number when the user unstaked the total balance from a delegator (used for 360 blocks cooldown) 

### 🖼️ **Screenshots:**


https://github.com/user-attachments/assets/ee0fac2a-90cb-4a4d-a395-866e4a26e21c

UrbanWill committed:
feat: display stake button for TAO token

UrbanWill committed:
feat: created bittensor validator hook, dinamically display pool info by chain id

UrbanWill committed:
feat: useGetStakingInfo hook, and auxiliary hooks to fetch staking for a given asset

UrbanWill committed:
feat: updated NoomPoolBondWizard, updated component types

UrbanWill committed:
feat: hooks for bittensor hotkeys and validators

UrbanWill committed:
feat: display bittensor validator name and unbound button

UrbanWill committed:
feat: hooks for fetching nomPooldata, consolidated data in  useGetUnbondInfo

UrbanWill committed:
feat: unbond TAO, totalStaked and unbondPayload hooks

UrbanWill committed:
fix: poolname, unbond btn and loading sekeleton displayed when it should not

UrbanWill committed:
fix: staking TAO with remark

UrbanWill committed:
chore: added taostats api key to env and webpack

UrbanWill committed:
feat: unbond TAO, display tao unbond info

UrbanWill committed:
feat: useGetBittensorMinJoinBond

UrbanWill committed:
refactor: conslidated data in useGetStakeInfo hook

UrbanWill committed:
fix: max button not working before input,
fetch payload with 0n plancks and invalidate form if pancks is < 0n

UrbanWill committed:
chore: tidy up useGetUnbondInfo

UrbanWill committed:
feat: added hooks to fetch and calculated if can stake bittensor

UrbanWill committed:
feat: display stake warning message and tooltip

UrbanWill committed:
chore: renamed NomPoolBond folder and items to "Bond"

UrbanWill committed:
chore: renamed NomPoolUnboond folder and items to "Unbond"

UrbanWill committed:
chore: created hooks folder, organized bittensor and nomPools hooks

UrbanWill committed:
fix: bittensor cooldown when switching accounts back and forth

UrbanWill committed:
chore: added comment

UrbanWill committed:
chore: renamed component name to be generic to bonds

UrbanWill committed:
feat: added BondSelectDrawer for popup UI

UrbanWill committed:
feat: added BittensorBondSelectDrawer

UrbanWill committed:
feat: hooks to fetch and merged Bittensor delegators data

UrbanWill committed:
feat: bondSelectDrawer, bond skeleton option

UrbanWill committed:
feat: sort delegators

UrbanWill committed:
feat: open modal with "select" option when asset is TAO

UrbanWill committed:
refactor: delegate select drawer to form in modal

UrbanWill committed:
fix: add 10 min staleTime for taostats queries

UrbanWill committed:
feat: hooks for fetching staked balance per delegator

UrbanWill committed:
fix: bondPoolName when doing anything but staking

UrbanWill committed:
refactor: display delegator stake position on AssetDetails components

UrbanWill committed:
fix: bondPooName should receive chainId as props

UrbanWill committed:
feat: unstake by delegator data

UrbanWill committed:
fix: stake cooldown when switching accounts

UrbanWill committed:
feat: track unstake block number and use it for stake cooldown

UrbanWill committed:
fix: bondOption token formartting, icon

UrbanWill committed:
chore: added changeset to util package

UrbanWill committed:
feat: fetch data from rpc instead of taostats, display TAO staked in all accounts

UrbanWill committed:
refactor: fetch total staked from rpc when unbonding, remove unused hooks

UrbanWill committed:
fix: delegators sorting race condition

UrbanWill committed:
chore: minor RemoteConfigStore type updates

UrbanWill committed:
chore: updated taostats api endpoint

UrbanWill committed:
chore: pr review change requests, added horizontal scrolling for sorting actions

UrbanWill committed:
chore: bond row label, bond form form grid spacing

UrbanWill committed:
feat: added taostats proxy api

alecdwm reviewed (dismissed):
👌

chidg commented:
Looking great, a few small things to add re: error handling of fetch requests.

When no `TAOSTATS_API_KEY` is configured, or else the `fetchBittensorInfiniteValidators` function fails for some reason, it would be good to display a nice message like "Unable to fetch validators" rather than just an empty screen

<img width="388" alt="Screenshot 2024-11-19 at 6 24 05 PM" src="https://github.com/user-attachments/assets/2531665b-e57f-4651-89ca-a60a6666b671">

chidg commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useGetBittensorSupportedDelegates.ts:
```diff
@@ -0,0 +1,22 @@
+const SUPPORTED_DELEGATES_URL =
+  "https://raw.githubusercontent.com/opentensor/bittensor-delegates/main/public/delegates.json"
+
+const fetchBittensorSupportedDelegates = async (): Promise<ValidatorsResponse> => {
```
Please handle failures of this request

chidg commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useGetBittensorValidator.ts:
```diff
@@ -0,0 +1,108 @@
+const fetchBittensorValidator = async (hotkey: string | undefined): Promise<BittensorValidator> => {
+  try {
+    return await (
+      await fetch(`${TAOSTATS_BASE_PATH}/api/validator/latest/v1?hotkey=${hotkey}`, {
```
This one does handle errors internally, but the error state doesn't seem like it's properly handled in the calling logic - see comment on `BondForm.tsx`

chidg commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useGetBittensorInfiniteValidators.ts:
```diff
@@ -0,0 +1,33 @@
+
+const fetchBittensorInfiniteValidators = async (page: number = 1): Promise<ValidatorsData> => {
+  return await (
+    await fetch(
```
Please handle errors in this request, or if the api key is not set

0xKheops commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useCombineBittensorStakeInfo.ts:
```diff
@@ -0,0 +1,123 @@
+  const tokenRates = useTokenRates(tokenId)
+  const selectedCurrency = useSelectedCurrency()
+
+  const totalStaked = balances.each.reduce((acc, b) => {
```
missing useMemo (checked with a console log and this hook is called a lot, even on non TAO token screens)
same with lines 29 and 42

0xKheops commented on apps/extension/src/ui/domains/Staking/Bond/BondForm.tsx:
```diff
@@ -24,26 +24,36 @@ import { Fiat } from "../../Asset/Fiat"
 
   if (!token) return null
 
+  const stakeAssetLabel = () => {
```
these labels need to be wrapped in t("") from here, or our localization parser wont pick them up

0xKheops commented on apps/extension/src/ui/domains/Staking/Bond/BondForm.tsx:
```diff
@@ -275,16 +286,56 @@ export const AmountEdit = () => {
+                <Tooltip>
+                  <TooltipTrigger asChild>
+                    <div className="text-alert-warn flex items-center justify-center gap-1 whitespace-nowrap text-xs">
+                      <InfoIcon className="inline-block" /> {t(stakeWarningMessage)}
```
t() should always contain the string itself, can't be variable or parser won't pick it up

0xKheops commented on apps/extension/src/ui/domains/Staking/Bond/useBondButton.ts:
```diff
@@ -20,14 +20,28 @@ export const useNomPoolBondButton = ({
+  const { open } = useBondModal()
+
+  // TODO: Add this to talisman-config
+  const mockedRemoteConfig = {
```
remove ?

UrbanWill replied:
Can remove when this gets an ✅  and merged: https://github.com/TalismanSociety/talisman-config/pull/7

0xKheops commented on apps/extension/src/ui/domains/Staking/Bond/useBondWizard.ts:
```diff
@@ -313,10 +250,11 @@ export const useNomPoolBondWizard = () => {
       })
 
     return null
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
any reason for this ?

0xKheops commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useBittensorStakedBlockNumber.ts:
```diff
@@ -0,0 +1,48 @@
+}
+
+export const useBittensorStakedBlockNumber = () => {
+  const [storedValue, setStoredValue] = useLocalStorage<StoredData | null>("initialValue", null)
```
Please provide a better name to this store, we need to be able to tell what it's for by it's name, when looking at all the storages entry of the domain.
Also not sure about the fact that we should use local storage like this at all, we usually implement our stores in extension-core with `StorageProvider` or one of it's derivatives.
@chidg wdyt ?

chidg replied:
ah yeah, should have spotted this, generally we should not use the normal window local storage api in the extension. Better to add a key to one of the existing stores (probably `app` in this case)

0xKheops commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useBittensorStakedBlockNumber.ts:
```diff
@@ -0,0 +1,48 @@
+  const [storedValue, setStoredValue] = useLocalStorage<StoredData | null>("initialValue", null)
+
+  // Method to get data by account and delegator
+  const getByAccountAndDelegator = ({
```
missing useCallback
same below

0xKheops commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useCanStakeBittensor.ts:
```diff
@@ -0,0 +1,52 @@
+
+  const { getByAccountAndDelegator } = useBittensorStakedBlockNumber()
+
+  const unbondBlock = getByAccountAndDelegator({ account: address, delegator: hotkey })
```
missing useMemo

0xKheops commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useGetBittensorValidator.ts:
```diff
@@ -0,0 +1,108 @@
+  hotkeys: (string | undefined)[]
+  isEnabled?: boolean
+}) => {
+  return useQueries({
```
That's very sexy ser :)

UrbanWill replied:
Thanks!
Edit: no, you are! ❤️

0xKheops commented on apps/extension/src/ui/domains/Staking/shared/BondPoolName.tsx:
```diff
@@ -0,0 +1,71 @@
+    poolName,
+    defaultPoolName = "Talisman Pool"
+
+  const hookMap = {
```
Please refactor with 2 child compoenents, selectively rendered depending on chainId
this will remove the need for the hook map and "let" variables in component body, for easier maintenance and readability
hook calls should never be conditional.

0xKheops commented on apps/extension/src/ui/domains/Staking/shared/StakingFeeEstimate.tsx:
```diff
@@ -18,7 +18,7 @@ export const StakingFeeEstimate: FC<{
           Failed to estimate fee
         </div>
-      ) : plancks && tokenId ? (
+      ) : (plancks || plancks === 0n) && tokenId ? (
```
a 0 fee would be a bug, how did this happen ?

UrbanWill replied:
That's actually the extrinsic fee for unstaking TAO, check this tx as an example https://taostats.io/extrinsic/4264248-0007?network=finney

0xKheops commented on apps/extension/src/ui/domains/Staking/shared/StakingUnbondingPeriod.tsx:
```diff
@@ -1,48 +1,36 @@
+    isLoading = false,
+    duration = t("N/A")
+
+  const hookMap = {
```
same as BondPoolName

0xKheops commented on apps/extension/src/ui/domains/Transactions/TxProgress.tsx:
```diff
@@ -118,6 +118,7 @@ const useTxStatusDetails = (tx?: WalletTransaction) => {
           animStatus: "failure",
         }
       case "success":
+        onSuccess?.(tx.blockNumber)
```
please move in it's own useEffect based on the single tx.status dependency, or this will be called multiple times
also i think this useEffect should be in TxProgressBase, as it's unrelated with tx details to be displayed

0xKheops commented on packages/util/src/formatTokenDecimals.ts:
```diff
@@ -0,0 +1,4 @@
+export const formatTokenDecimals = (amount: number | string, decimals: number): number => {
```
please remove and use plancksToTokens instead
this one seems unsafe, notably it looks like it could output more decimals than the token supports

0xKheops reviewed (changes requested):
some minor things, please check below.

another that didnt fit to a particular part of the code is that i see bittensor staking related hooks being called a lot for no reason, especially in AssetDetails
calling any of them loads bittensor sapi object which downloads the full metadata.
Unless a bittensor balance is detected, it would be nice to make sure those hooks do nothing. 
Note: at this time even if if bittensor is disabled in settings, they will fetch metadata and execute all the logic

chidg commented on apps/extension/src/ui/domains/Staking/Bond/BondForm.tsx:
```diff
@@ -275,16 +286,56 @@ export const AmountEdit = () => {
+
+  switch (token?.chain?.id) {
+    case "bittensor":
+      ;({ data, isLoading } = hookMap["bittensor"](poolId as unknown as string))
```
Not such a big fan of this syntax for this, but either way would prefer to include the `isError`state here so we can show an appropriate message

chidg commented on apps/extension/src/ui/domains/Staking/shared/BondPoolName.tsx:
```diff
@@ -0,0 +1,71 @@
+    case "bittensor":
+      ;({ data, isLoading, isError } = hookMap["bittensor"](poolId as unknown as string))
+      poolName = data?.data?.[0].name || ""
+      poolName = (
```
Duplicate variable assignment, is that intentional?

chidg commented on apps/extension/src/ui/domains/Staking/shared/useGetLatestBlockNumber.ts:
```diff
@@ -0,0 +1,18 @@
+
+export const useGetLatestBlockNumber = ({ sapi, isEnabled }: GetLatestBlockNumber) => {
+  return useQuery({
+    queryKey: ["useGetLatestBlockNumber", sapi?.id],
```
I got an error on this line when testing (no Taostats API key configured) - `Query data cannot be undefined. Please make sure to return a value other than undefined from your query function. Affected query key: ["useGetLatestBlockNumber",null]`

chidg reviewed (changes requested):
Few more comments, apologies - these got lost in vs code PR review process somewhere

UrbanWill committed:
feat: handle missing origin/auth headers

UrbanWill committed:
refactor: display UI errors

UrbanWill committed:
refactor: error handling

UrbanWill committed:
refactor: apr error msg, duplicated var assignement, null query key error

UrbanWill committed:
perf: memoized values and disabled queries when token is not tao or tao bal is 0

UrbanWill committed:
chore: memoized variables, fixed transalation, removed eslint comment

UrbanWill committed:
refactor: broke down BondPoolName into multiple child components

UrbanWill committed:
refactor: broke down StakingUnbondingPeriod into multiple child components

UrbanWill committed:
fix: only display error msg in BondOption if there is actually an error

UrbanWill committed:
refactor: out with formatTokenDecials in planckToTokens

UrbanWill committed:
refactor: use minJoinBond to get an accurate fake fee estimation

UrbanWill committed:
refactor: moved onSuccess ot its opwn useEffect in TxProgressBase

UrbanWill committed:
chore: removed mocked remote config

UrbanWill committed:
chore: removed redundant translation, early bittensor delegator name button return

UrbanWill committed:
chore: early bittensor delegator name button return, cath api error

0xKheops commented on apps/extension/src/ui/domains/Transactions/TxProgress.tsx:
```diff
@@ -241,14 +256,21 @@ const TxProgressEvm: FC<TxProgressEvmProps> = ({ tx, className, onClose }) => {
+  networkIdOrHash,
+  onClose,
+  className,
+  onSuccess,
```
Sorry i didnt dig this thing enough last time. 
It feels like success detection shouldnt originate from here, which is a UI component that is meant to support both EVM and substrate.

Option 1 : use useTransaction(hash) from useUnbondWizard() to detect success from there, without needing any callbacks, and rollback all changes to this component
Option 2 : move the success check from TxProgressSubstrate up to TxProgress, so it can support both substrate and evm

Preference for option 1 though

UrbanWill replied:
Went with option um, thanks for the suggestion!

0xKheops reviewed (dismissed):
Sorry, one more change needed in TxProgress :)

UrbanWill committed:
refactor: handle unbond success on useUnbondWizard, revert TxProgress changes

UrbanWill committed:
refactor: create store and custom hooks for keeping track of bittensor unbond block numbers

UrbanWill committed:
refactor: change bittensor staking flow,
default poolId to the currently staked delegator, allow user delegator selection

UrbanWill committed:
feat: place recommented on top of delegators list

UrbanWill committed:
feat: recommended delegator loading skeleton

UrbanWill committed:
refactor: use app store to track usntaking tx block number

chidg commented on apps/extension/src/ui/domains/Staking/hooks/bittensor/useCanStakeBittensor.ts:
```diff
@@ -0,0 +1,59 @@
+      Number(unbondBlockNumber ?? 0),
+      Number(lastStakedBlockNumber ?? 0),
+    )
+    const coolDownBlockNumber = Number(latestActionBlock) + STAKE_INTERVAL_BLOCKS
```
```suggestion
    const coolDownBlockNumber = latestActionBlock
      ? Number(latestActionBlock) + STAKE_INTERVAL_BLOCKS
      : 0
```

chidg replied:
This prevents a flash of the `canStake === false` state when the screen first loads

UrbanWill committed:
revert: uncommented code

UrbanWill committed:
style: changed color of non-blocking action messages

UrbanWill committed:
fix: center tooltip info icon, wait for blocknumber in stake cooldown warning on first load

UrbanWill committed:
refactor: remove redundand fallback for blockNumber

chidg committed:
chore: fix text copy

UrbanWill committed:
refactor: updated BondOption design

UrbanWill committed:
chore: pulled latest changes from remote

UrbanWill committed:
style: minor spacing and color adjustments

UrbanWill committed:
Merge branch 'dev' of github.com:TalismanSociety/talisman into feat/tao-staking

UrbanWill committed:
chore: rename "direct staking" => "delegated staking"

chidg committed:
chore: add separator after recommended validator in bittensor validator selection

UrbanWill committed:
chore: ran pnpm changeset

chidg committed:
fix: nom pool staking warning copy

UrbanWill committed:
Merge branch 'feat/tao-staking' of github.com:TalismanSociety/talisman into feat/tao-staking

chidg reviewed (dismissed):
LGTM!

chidg committed:
fix: canary build icon names

UrbanWill committed:
chore: updated copy to "featured"

UrbanWill committed:
Merge branch 'feat/tao-staking' of github.com:TalismanSociety/talisman into feat/tao-staking

chidg reviewed (approved):
LGGGGTM
